[inductor] support linear & layer_norm unbacked

ColinPeppler opened the pull request:
### What
- Use `statically_known_true` over `guard_size_oblivious` in cases where we're checking an optimization path. Otherwise, it will DDE and we can't take the safe/slower path.
- For broadcast checks, use `fallback=False` if we encounter a DDE. Typically, unbackeds would be ≥2 and that falls inline with size-oblivious reasoning (i.e. when `size_oblivious=True`).

### Example DDE
```
torch._inductor.exc.InductorError: LoweringException: GuardOnDataDependentSymNode: Could not guard on data-dependent expression Eq((u0//387), 1) (unhinted: Eq((u0//387), 1)).  (Size-like symbols: u0)

Caused by: (_inductor/lowering.py:488 in broadcast_symbolic_shapes)
```
```
torch._inductor.exc.InductorError: LoweringException: GuardOnDataDependentSymNode: Could not guard on data-dependent expression Eq((u0//387), 1) (unhinted: Eq((u0//387), 1)).  (Size-like symbols: u0)

Caused by: (_inductor/ir.py:2797 in create)
```

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #155267



cc @ezyang @penguinwu @bobrenjc93 @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben

ColinPeppler committed:
[inductor] support linear & layer_norm unbacked

[ghstack-poisoned]

eellison reviewed (commented):
cc @zou3519 , @bdhirsh - what's our policy for this ? How do we reconcile "strides wrong is a ubn" and now introducing meta functions where we are potentially not computing strides correctly.

zou3519 commented:
> cc @zou3519 , @bdhirsh - what's our policy for this ? How do we reconcile "strides wrong is a ubn" and now introducing meta functions where we are potentially not computing strides correctly.

We discussed this at composability sync. The conclusion was:
1) strides wrong for backed symints is a UBN
2) unbacked symints are explicitly wrong. There's no way to fulfill the requirements for unbacked symints and have correct strides at the same time. So unbacked symints being "wrong" is not a problem.

ColinPeppler committed:
Update on "[inductor] support linear & layer_norm unbacked"

### What
- Use `statically_known_true` over `guard_size_oblivious` in cases where we're checking an optimization path. Otherwise, it will DDE and we can't take the safe/slower path.
- For broadcast checks, use fallback=False if we encounter a DDE. Typically, unbackeds would be ≥2 and that falls inline with size-oblivious reasoning.
- In `coerce_tangent_and_suggest_memory_format`, use the torch._refs version of `contiguous` over Aten's version because Aten doesn't have a DDE-friendly path (yet).

### Example DDE
```
GuardOnDataDependentSymNode: Could not guard on data-dependent expression Eq(128*((u0//387)), 0) (unhinted: Eq(128*((u0//387)), 0)).  (Size-like symbols: u0)

Caused by: (_functorch/_aot_autograd/collect_metadata_analysis.py:84 in coerce_tangent_and_suggest_memory_format)
```

```
torch._inductor.exc.InductorError: LoweringException: GuardOnDataDependentSymNode: Could not guard on data-dependent expression Eq((u0//387), 1) (unhinted: Eq((u0//387), 1)).  (Size-like symbols: u0)

Caused by: (_inductor/lowering.py:488 in broadcast_symbolic_shapes)
```
```
torch._inductor.exc.InductorError: LoweringException: GuardOnDataDependentSymNode: Could not guard on data-dependent expression Eq((u0//387), 1) (unhinted: Eq((u0//387), 1)).  (Size-like symbols: u0)

Caused by: (_inductor/ir.py:2797 in create)
```




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

ColinPeppler committed:
Update on "[inductor] support linear & layer_norm unbacked"

### What
- Use `statically_known_true` over `guard_size_oblivious` in cases where we're checking an optimization path. Otherwise, it will DDE and we can't take the safe/slower path.
- For broadcast checks, use fallback=False if we encounter a DDE. Typically, unbackeds would be ≥2 and that falls inline with size-oblivious reasoning.
- In `coerce_tangent_and_suggest_memory_format`, use the torch._refs version of `contiguous` over Aten's version because Aten doesn't have a DDE-friendly path (yet).

### Example DDE
```
GuardOnDataDependentSymNode: Could not guard on data-dependent expression Eq(128*((u0//387)), 0) (unhinted: Eq(128*((u0//387)), 0)).  (Size-like symbols: u0)

Caused by: (_functorch/_aot_autograd/collect_metadata_analysis.py:84 in coerce_tangent_and_suggest_memory_format)
```

```
torch._inductor.exc.InductorError: LoweringException: GuardOnDataDependentSymNode: Could not guard on data-dependent expression Eq((u0//387), 1) (unhinted: Eq((u0//387), 1)).  (Size-like symbols: u0)

Caused by: (_inductor/lowering.py:488 in broadcast_symbolic_shapes)
```
```
torch._inductor.exc.InductorError: LoweringException: GuardOnDataDependentSymNode: Could not guard on data-dependent expression Eq((u0//387), 1) (unhinted: Eq((u0//387), 1)).  (Size-like symbols: u0)

Caused by: (_inductor/ir.py:2797 in create)
```




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

laithsakka commented on torch/_inductor/lowering.py:
```diff
@@ -943,18 +943,18 @@ def broadcast_tensors(*inputs):
                 (
                     V.graph.sizevars.shape_env.evaluate_expr(
-                        sympy.Eq(a, 1), size_oblivious=True
+                        sympy.Eq(a, 1), size_oblivious=True, fallback_value=False
```
remove size_oblivious=True

ColinPeppler committed:
Update on "[inductor] support linear & layer_norm unbacked"

### What
- Use `statically_known_true` over `guard_size_oblivious` in cases where we're checking an optimization path. Otherwise, it will DDE and we can't take the safe/slower path.
- For broadcast checks, use fallback=False if we encounter a DDE. Typically, unbackeds would be ≥2 and that falls inline with size-oblivious reasoning.

### Example DDE
```
torch._inductor.exc.InductorError: LoweringException: GuardOnDataDependentSymNode: Could not guard on data-dependent expression Eq((u0//387), 1) (unhinted: Eq((u0//387), 1)).  (Size-like symbols: u0)

Caused by: (_inductor/lowering.py:488 in broadcast_symbolic_shapes)
```
```
torch._inductor.exc.InductorError: LoweringException: GuardOnDataDependentSymNode: Could not guard on data-dependent expression Eq((u0//387), 1) (unhinted: Eq((u0//387), 1)).  (Size-like symbols: u0)

Caused by: (_inductor/ir.py:2797 in create)
```




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

eellison commented on torch/_inductor/lowering.py:
```diff
@@ -943,18 +943,18 @@ def broadcast_tensors(*inputs):
-                        sympy.Eq(a, 1), size_oblivious=True
+                        sympy.Eq(a, 1), fallback_value=False
                     )
                     and not V.graph.sizevars.shape_env.evaluate_expr(
-                        sympy.Eq(b, 1), size_oblivious=True
+                        sympy.Eq(b, 1), fallback_value=False
                     )
                 )
                 or (
                     not V.graph.sizevars.shape_env.evaluate_expr(
-                        sympy.Eq(a, 1), size_oblivious=True
+                        sympy.Eq(a, 1), fallback_value=False
                     )
                     and V.graph.sizevars.shape_env.evaluate_expr(
-                        sympy.Eq(b, 1), size_oblivious=True
+                        sympy.Eq(b, 1), fallback_value=False
                     )
```
Can we reuse the above `broadcast_symbolic_shapes` here instead of duplicating the logic ? (i know this was preexisting to your change)

ColinPeppler replied:
this one would look a lil bit forced since `broadcast_symbolic_shapes` returns the broadcasted shape vs. here we just want to check for broadcast. 

But agreed the code looks duplicated, I'm gonna find a more reusable way.

eellison reviewed (approved):
Looks reasonable

ColinPeppler committed:
Update on "[inductor] support linear & layer_norm unbacked"

### What
- Use `statically_known_true` over `guard_size_oblivious` in cases where we're checking an optimization path. Otherwise, it will DDE and we can't take the safe/slower path.
- For broadcast checks, use `fallback=False` if we encounter a DDE. Typically, unbackeds would be ≥2 and that falls inline with size-oblivious reasoning (i.e. when `size_oblivious=True`).

### Example DDE
```
torch._inductor.exc.InductorError: LoweringException: GuardOnDataDependentSymNode: Could not guard on data-dependent expression Eq((u0//387), 1) (unhinted: Eq((u0//387), 1)).  (Size-like symbols: u0)

Caused by: (_inductor/lowering.py:488 in broadcast_symbolic_shapes)
```
```
torch._inductor.exc.InductorError: LoweringException: GuardOnDataDependentSymNode: Could not guard on data-dependent expression Eq((u0//387), 1) (unhinted: Eq((u0//387), 1)).  (Size-like symbols: u0)

Caused by: (_inductor/ir.py:2797 in create)
```




cc ezyang penguinwu bobrenjc93 voznesenskym EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

[ghstack-poisoned]
